$OpenBSD: patch-lib_zip_name_locate_c,v 1.1.2.1 2011/03/22 09:24:40 jasper Exp $

Security fix for CVE-2011-0421
libzip "_zip_name_locate()" NULL Pointer Dereference Vulnerability

Patch from upstream mercurial:
http://hg.nih.at/libzip/?cs=13654bfdc88c

--- lib/zip_name_locate.c.orig	Sun Jan 31 23:00:30 2010
+++ lib/zip_name_locate.c	Tue Mar 22 10:21:14 2011
@@ -1,6 +1,6 @@
 /*
   zip_name_locate.c -- get index by name
-  Copyright (C) 1999-2007 Dieter Baron and Thomas Klausner
+  Copyright (C) 1999-2011 Dieter Baron and Thomas Klausner
 
   This file is part of libzip, a library to manipulate ZIP archives.
   The authors can be contacted at <libzip@nih.at>
@@ -59,7 +59,12 @@ _zip_name_locate(struct zip *za, const char *fname, in
 	_zip_error_set(error, ZIP_ER_INVAL, 0);
 	return -1;
     }
-    
+
+    if ((flags & ZIP_FL_UNCHANGED)  && za->cdir == NULL) {
+        _zip_error_set(error, ZIP_ER_NOENT, 0);
+        return -1;
+    }
+
     cmp = (flags & ZIP_FL_NOCASE) ? strcasecmp : strcmp;
 
     n = (flags & ZIP_FL_UNCHANGED) ? za->cdir->nentry : za->nentry;
